### PR TITLE
Forbidden CSP directives

### PR DESCRIPTION
**Pull request by atropnikov**
Forbidden CSP directives. #685

### CODE REVIEW

**Comment by atropnikov** [comment]
These limitations for a rule's creator.
1. So `,` will break rule parsing at all, also as `$` sign.
2. I've just realized, that I completely forgot about `subdocument` and `important` modifiers. I'll add processing for these ones in this pull.

File: Extension/lib/filter/rules/csp-filter.js [hunk 1]
```diff
@@ -93,17 +99,30 @@
...
-            var cspRules = [];
+            var blockingRules = cspBlockFilter.findRules(url, documentHost, thirdParty, requestType);
 
-            var blockingRules = cspBlockFilter.findRules(url, documentHost, thirdParty, adguard.RequestTypes.CSP);
+            var cspRules = [];
 
             // Collect whitelist and blocking CSP rules in one array
             if (blockingRules) {
+
+                // Add important rules, ignore whitelist
```
Comment (Mizzick) [inline Extension/lib/filter/rules/csp-filter.js]: Why not to collect it in one iteration?

File: Extension/lib/filter/rules/csp-filter.js [hunk 2]
```diff
@@ -72,11 +72,17 @@
...
          * @param thirdParty true if request is third-party
+         * @param requestType   Request content type. $CSP rules are applied only at DOCUMENT or SUB_DOCUMENT levels.
          * @returns Matching rules
          */
-        function findCspRules(url, documentHost, thirdParty) {
+        function findCspRules(url, documentHost, thirdParty, requestType) {
 
-            var whiteRules = cspWhiteFilter.findRules(url, documentHost, thirdParty, adguard.RequestTypes.CSP);
+            /**
+             * DOCUMENT request type at the rule's level has other meaning, so replace it with OTHER modifier
```
Comment (Mizzick) [inline Extension/lib/filter/rules/csp-filter.js]: Please add more informative comment, I don't understand why to replace it?

File: Extension/lib/filter/rules/url-filter-rule.js [hunk 3]
```diff
@@ -646,7 +657,6 @@
     UrlFilterRule.contentTypes.ALL |= UrlFilterRule.contentTypes.FONT;
     UrlFilterRule.contentTypes.ALL |= UrlFilterRule.contentTypes.WEBSOCKET;
```
Comment (Mizzick) [inline Extension/lib/filter/rules/url-filter-rule.js]: Do we need to have websockets and webrtc here aswell?

Comment (atropnikov) [inline Extension/lib/filter/rules/url-filter-rule.js]: Why not? `WEBSOCKET` is a request type, just like `IMAGE` or `SCRIPT`.
`WEBRTC` isn't supported by browsers now, but it's possible in the future.

File: Extension/lib/filter/rules/csp-filter.js [hunk 4]
```diff
@@ -19,6 +19,51 @@
...
+    function putWithPriority(rule, whiteListRule, map) {
+
+        var cspDirective = rule.cspDirective;
+        var existRule = map[cspDirective];
+
+        var pr1 = getRulePriority(rule);
+        var pr2 = getRulePriority(whiteListRule);
+        var pr3 = getRulePriority(existRule);
+
+        var max = Math.max(pr1, pr2, pr3);
```
Comment (Mizzick) [inline Extension/lib/filter/rules/csp-filter.js]: How it will be if for example `pr1 == pr3`?

Comment (atropnikov) [inline Extension/lib/filter/rules/csp-filter.js]: If `pr1 === pr3` then rule1 is the same as rule3, so it doesn't matter which rule is used in the result.

**Pull request merged**